### DISCUSSION TITLE
Workflow: UV texture map generation with ControlNet Image Segmentation

### DISCUSSION THREAD

**Discussion by AugmentedRealityCat** [Mikubill/sd-webui-controlnet#204] [Show and tell]
Here is how to use ControlNet with the Image Segmentation model to generate unwrapped UV texture maps for your 3d objects.

This video shows 50 different packaging textures for this box, all synthesized in a few minutes. Textures were applied in Cinema4d, and the model used for this example was taken from its assets library, but this should work with any 3d application.

https://user-images.githubusercontent.com/114955776/219891976-15219bf5-4668-4b4f-98f0-0d913b9bc9fe.mp4

**Workflow**

1. Create or import your model and unwrap its UV coordinates. It is essential to have clean UVs for this to work.

2. Create a segmentation map by applying flat plain colors to the different surfaces of your object. Keep it simple. Reuse the same color for surfaces sharing the same properties.
![workflow_UV_mapping](https://user-images.githubusercontent.com/114955776/219892249-2803555f-7d7c-4b43-83d0-e5f54a00c9d9.png)

3. Export that segmentation map as a PNG. It does not have to be square but it's a good idea to have it in a resolution that works well with SD. For this example, it was in 2048x1024, a nice 2:1 ratio that you can downscale easily to 1024x512 if required.
![unwrappedbox_v2](https://user-images.githubusercontent.com/114955776/219892330-6a691a40-c024-4588-b044-2a2d16cf9dfc.png)

4. Start Stable Diffusion and enable the ControlNet extension.

5. Load your segmentation map as an input for ControlNet. Make sure you set the resolution to match the ratio of the texture you want to synthesize. I went for half-resolution here, with 1024x512.

6. Leave the Preprocessor to `None`. Since we already created our own segmentation map there is no need to reprocess it.

7. Set the ControlNet Model to the one containing the word *Seg* (for image SEGmentation) 

8. You may have to adjust the weight parameter - in this example I cranked it up to 2.

9. . Write a simple prompt to describe the texture you want to generate, for example Grape Juice Packaging. You can also change the sampler, the steps, the CFG and the SD model. For this example I used the plain ema-only 1.5 model just to show you don't need anything fancy. You can also use other extensions, like in this case I wanted the texture to be seamless on the X axis, so I used the Asymmetric Tiling extension to achieve that.
![Stable Diffusion(1)](https://user-images.githubusercontent.com/114955776/219893894-4aa6c80d-91eb-4f9c-9d20-6b6fcc42bed1.jpg)

10.  Optional: if you want you can do this in IMG2IMG mode and use a picture to influence the texture map generation process. 
![Stable Diffusion(3)](https://user-images.githubusercontent.com/114955776/219893922-dbde9249-e5f3-4012-9af3-509b5c7170b0.jpg)

11. Optional: if you want to synthesize many variations, it can be a good idea to use wildcards and other dynamic prompting methods.

12. Press the Generate button, repeat until you get an image that satisfies you. Here is one example of the results I got. I made hundreds of them over a few minutes by using `__fruit__ juice packaging` as a wildcard prompt
![32088-3351088082-grape juice packaging](https://user-images.githubusercontent.com/114955776/219894163-583e59fc-c2e3-493f-bd1c-a429f4d34d63.png)

13. Import that picture as a texture map and apply it to your 3d object using the same UV coordinates as those used to first generate the Image Segmentation map.
![workflow_texture_applied](https://user-images.githubusercontent.com/114955776/219894013-0166ea11-437c-4a00-ae6a-62c4cc6808a7.png)

14. Press render and voila !

**Comment by VladCorvi** [comment]
Thank you!

**Comment by oliverban** [comment]
Holy shit.

**Reply by AugmentedRealityCat** [reply]
That's exactly what I said out loud when I synthesized the first of those packaging textures - the orange juice one made with IMG2IMG, which by the way doesn't even tile properly because I had yet to activate the asymmetric tiling at that point !

**Comment by Jongulo** [comment]
![image](https://user-images.githubusercontent.com/111000352/220204588-4d01272e-fbfc-4bc0-a9e8-dcf6eb4383f9.png)
impressive results, though manual inpainting was still required for a part.

**Reply by utsang** [reply]
How did u export the uv textures in blender, like how did you make them segmented into different colors for each part?

**Comment by FernandoConti** [comment]
This is marvelous and delicious! And I'm not talking about those gorgeous packagings haha!

This is the type of SD application I want to see more of, please share your new ideas! 🥇🥇🥇🥇🥇

**Comment by deloyed** [comment]
Wow, incredible

**Comment by p0mad** [comment]
@AugmentedRealityCat
How did to make a video out?
I'm not using the same model, but using the openpose face and i want to input a video of 3Seconds of human pose and get the video output of the generated human!

Would you please help me on that?
Thanks
best regards

**Reply by AugmentedRealityCat** [reply]
Video was rendered from Cinema4d - only the textures were synthesized with Stable Diffusion + ControlNet Image Segmentation

**Reply by p0mad** [reply]
@AugmentedRealityCat 
Great, So You have not used SDControlnet for creating initial Video for cinema4D, it was just Images. Right?

**Reply by AugmentedRealityCat** [reply]
Yes. Stable Diffusion was used just to generate the 2d texture maps that were then applied on the 3d models in a different software, Cinema4d, and that's also where I made the animation with the box turning and the texture changing by going through all the packaging variations I had generated earlier.

ControlNet's Semantic Segmentation  was used to make sure the packaging images generated with Stable Diffusion would fit perfectly with the 3d model of the box afterwards. The texture generated with Stable Diffusion is then basically wrapped around the 3d model using a set of correspondance between each polygon on your model (XYZ coordinates for each vertex) and the part of your 2d texture that should be used to "paint" that polygon (the UV coordinates for each vertex).

**Comment by LordLiang** [comment]
Wonderful! Are there more complex examples? For example, a car mesh or an animal mesh?

**Comment by turkyden** [comment]
@AugmentedRealityCat Can you provide me with the .gltf model file of this packaging box? Thanks ！！!

**Reply by AugmentedRealityCat** [reply]
The packaging model and the corresponding mapping reference are assets that are provided as part of an online library that is accessible when you subscribe to a Cinema4d license. 
Is there any way I can send that to you personally ? I can't redistribute those assets online by posting a link here, but I can share them personally with you, for academic research purposes.

**Comment by anwoflow** [comment]
honestly this .. shouldn't work :) 
the segmentation model was trained on specific classes, so this basically is limiting the individual UV derived areas to 'a thing', but it's not going to produce semantically meaningful results - tires on a car in the right place of the UVs, etc. 

If it does  I will have to revisit my whole knowledge of this space and go into hiding to meditate.

**Reply by anwoflow** [reply]
but to be clear that's one clever idea, and I believe a model trained on a set of models with matching UVs could actually work.

**Reply by AugmentedRealityCat** [reply]
> If it does I will have to revisit my whole knowledge of this space and go into hiding to meditate.

It actually does work ! The thing is that when I first began playing with this there was almost no information available about that model and how to use it. So I tried different things, but I could not find any color reference chart that would tell me if that or this color was meant to represent a house or a car. And it's by experimenting that I discovered this "feature".

I now understand how it all works - I made color reference charts for semantic segmentations models that I shared over here just a few weeks after this was posted, and wrote a few tutorial explaining how to use those. 

But if I had known all of that then, I probably would have never tried it, and I would have never discovered it could be used like this.

Try it, you'll see, it works even though it should not, as we both know now. Ignorance is bliss sometimes ! 

No need to go hide anywhere and meditate, just laugh about my stupidity, and take advantage of what I found. Maybe you'll discover some other tricks to bend those semantic segmentation models even further.

**Reply by NicolaiLolansen** [reply]
What does the prompting do then? Is the prompting only to supply auxiliary information about the textures generated, like art style or similar? If the segmentation colors mean one thing, and you prompt for something else entirely, then what happens?

**Comment by NicolaiLolansen** [comment]
Is it possible to somehow influence what parts of the segmentation map the prompt applies to? Ie. redish colors are brick, blue is grass.

**Reply by AugmentedRealityCat** [reply]
When I made this I had no idea, but I can now tell you that, no, you can not influence what color is refering to what subject. 

The reason is that there is a pre-established semantic color chart for this model, and I had no idea about this either at the time ! Since that information was not readily available, I made and shared such color charts over here. Basically, you have a color to tell that this part is the sky, this part the road, and this other part a car. There are other semantic segmentation models available, and each has its own color reference system.

Look at the discussion page and you should find the semantic segmentation color chart that will tell you what color corresponds to grass.

An alternative is to use a photo of grass as an input for the pre-processor and pick the color that was used to cover the grassy areas. That's how I made the first chart before someone helped me find the part of the code that was responsible for that.

**Reply by ljleb** [reply]
For what it's worth, if you have a preprocessed color map with all wrong colors and you want to fix the colors, you can always remap which colors correspond to what concept before sending to controlnet using a fixed color mapping where i.e. "reddish" maps to the expected semantic color for bricks.

I wonder what happens if you pick in-between colors that were not explicitly trained? If that is useful somehow with the original model, you might not be able to interpolate in the same way depending on the color mapping you settle on.

**Reply by sdbds** [reply]
> For what it's worth, if you have a preprocessed color map with all wrong colors and you want to fix the colors, you can always remap which colors correspond to what concept before sending to controlnet using a fixed color mapping where i.e. "reddish" maps to the expected semantic color for bricks.
> 
> I wonder what happens if you pick in-between colors that were not explicitly trained? If that is useful somehow with the original model, you might not be able to interpolate in the same way depending on the color mapping you settle on.

Speaking from my experience in training anime face segment, if a corresponding color appears, it will be referenced as a conditional input similar to a I2I.

**Reply by ljleb** [reply]
I'm not sure I understand what you mean, sorry. To be clear, what I said above is a consequence of the idea that in the end, what matters is that the colors the model receives are correct with respect to training colors. The way you create a proper segmentation map really does not matter. So if you want to remap the colors, you just need a preprocessing step where you remap the colors to what they should be.

**Reply by NicolaiLolansen** [reply]
Okay, thanks for the replies. I found an extension to Stable Diffusion that uses multiple diffusion for different regions of an image, but geometrized. If one could take this idea, and apply it not to a square geometric region, but to a mask that's created by a color, then one could achieve stand-alone multi class prompting. You would then have to define a layer order, and merge the different output layers into a single image. 

It's quite limiting that the classes for the segmentation are pre-defined if you want to create UV textures I think. Thanks for the input though!

**Reply by AugmentedRealityCat** [reply]
> I wonder what happens if you pick in-between colors that were not explicitly trained?

The in-between colors are actually used to separate objects that share the same semantic class. If you have a single car in your image, then it will get the "official" semantic color for cars. But if you have 5 cars at the same time, then each car will get a slight variation of that official color, This allows the system to treat each car as its own object.

**Reply by AugmentedRealityCat** [reply]
> I found an extension to Stable Diffusion that uses multiple diffusion for different regions of an image, but geometrized. If one could take this idea, and apply it not to a square geometric region,

I think you will like this extension by @continue-revolution then ! 
https://github.com/continue-revolution/sd-webui-segment-anything 

And this one from @Uminosachi !
https://github.com/Uminosachi/sd-webui-inpaint-anything

**Reply by NicolaiLolansen** [reply]
Thank you @AugmentedRealityCat - I will check it out. Appreciate the replies

**Reply by elisabeth-jars** [reply]
Here is the table for what colors correspond to which class: 

ufade20k: https://docs.google.com/spreadsheets/d/1se8YEtb2detS7OuPE86fXGyD269pMycAWe2mtKUj2W8/edit?gid=0#gid=0 
ADE20k: https://docs.google.com/spreadsheets/d/1se8YEtb2detS7OuPE86fXGyD269pMycAWe2mtKUj2W8/edit#gid=0

From the segmentation section of https://stable-diffusion-art.com/controlnet/

**Comment by robiot** [comment]
Where can I find the seg difference model? I can't really find it anywhere.

**Reply by robiot** [reply]
https://huggingface.co/lllyasviel/ControlNet-v1-1/tree/main
`control_v11p_sd15_seg.pth`

I think I found it
太酷了
